pxtcompiler/emitter/transpile.ts

```diff
@@ -0,0 +1,115 @@
+// TODO: enable reference so we don't need to use: (pxt as any).py
+//      the issue is that this creates a circular dependency. This
+//      is easily handled if we used proper TS modules.
+//// <reference path="../../built/pxtpy.d.ts"/>
+
+// This code centralizes transpiling between typescript and python
+// so that we can cache results for better performance and user
+// experience (since compile -> decompile round-trips are lossy)
+namespace ts.pxtc.transpile {
+    export interface TranspileResult {
+        diagnostics: pxtc.KsDiagnostic[],
+        success: boolean,
+        outfiles: { [key: string]: string }
+    }
+
+    const mainName = (l: Lang) => `main.${l}`
+
+    export type Lang = "py" | "blocks" | "ts"
+    export interface LangEquivSet {
+        comparable: { [key in Lang]: number },
+        code: { [key in Lang]: string },
+    }
+    // a circular buffer of size MAX_CODE_EQUIVS that stores
+    // sets of equivalent code files so that when we translate
+    // from ts -> py -> ts we get back the same ts.
+    let codeEquivalences: LangEquivSet[] = []
```

Comment:
when is the cache cleared? it should be cleared whenever we change header.

Reply:
Actually, should be changed whenever dependencies change as well. Or maybe a hashed version of config.json should be stored with the text.

Reply:
Don't use hashes unless necessary - storing a string has essentially no cost (just a pointer) and comparison is cheap.

Reply:
Okay, so I'm using the code + all the qNames as the key (no hash). Does that work for you @pelikhan + @riknoll ?

Reply:
Extensions can modify existing qNames without adding any new ones, we do that in the Minecraft target for example. Probably better to use all of the extension names and their installed versions.

Reply:
Ah okay I'll use pxt.json and/or config.json

Reply:
pxt.json isn't accessible in many of the places that call the transpilers like service.ts as far as I can tell

Using the filenames seems to capture pretty well which packages are installed but not their versions:
![image](https://user-images.githubusercontent.com/6453828/70158222-bf004880-166b-11ea-872b-fa7945c448a6.png)

@mmoskal , do you know of a way to uniquely identify the versions of packages used from the CompileOptions? that would be the most convenient.

---
Realistically, it should be pretty rare that changing a API could cause different ts<->py transpilation. The transpilation is mostly syntactic. I can't think of any way an API could change that would keep the same .ts but different .py for example.

Reply:
Peli and I talked about this offline. I'm going to invalidate the cache based on the event because it's too difficult to do it the other way and because that's how everything else is done.

Reply:
Alright, I pushed the change where we invalidate in the service.ts operations.reset handler. Caching still works and if you open a new tab I can see the cache is reset.

I think this PR is ready. @pelikhan ?